README.md
Debian-PHP5
===========

Debian Package for PHP5

debian/NEWS
php5 (5.3.3-7+squeeze5) squeeze-security; urgency=high

  * The following new directives were added as part of security fixes:
    - max_input_vars - specifies how many GET/POST/COOKIE input variables
      may be accepted.  Default value is set to 1000.
    - xsl.security_prefs - define forbidden operations within XSLT
      stylesheets.  Write operations are now disabled by default.

 -- Ondřej Surý <ondrej@debian.org>  Mon, 23 Jan 2012 12:22:26 +0100

php5 (5.3.3-7+squeeze4) squeeze-security; urgency=low

  * Updated blowfish crypt() algorithm fixes the 8-bit character handling
    vulnerability (CVE-2011-2483) and adds more self-tests.  Unfortunately
    this change is incompatible with some old (wrong) generated hashes for
    passwords containing 8-bit characters.  Therefore the new salt prefix
    '$2x$' was introduced which can be used as a replacement for '$2a$'
    salt prefix in the password database in case the incompatibility is
    found.

 -- Ondřej Surý <ondrej@debian.org>  Mon, 04 Jul 2011 10:31:16 +0200

php5 (5.3.1-3) unstable; urgency=low

  * mod_php disabled in userdirs.

  The default Debian libapache2-mod-php5 package now disables the PHP
  engine on ~/public_html directories when mod_userdir is enabled, for
  security reasons.  Although discouraged, it can be re-enabled by
  commenting the <IfModule mod_userdir.c> block in
  /etc/apache2/mods-available/php5.conf

  * PHP 5.2 compatibility settings

  Given the short time to the Squeeze release freeze, the
  short_open_tag setting has been turned On again (upstream now
  defaults to Off on the php.ini files.) However, the request_order and
  auto_globals_jit settings continue to be the default from upstream
  ("GP" and On, respectively.)

 -- Raphael Geissert <geissert@debian.org>  Mon, 11 Jan 2010 16:49:28 -0600

php5 (5.2.11.dfsg.1-2) unstable; urgency=high

  * Maximum number of file uploads per request limited

  To prevent Denial of Service attacks by exhausting the number of
  available temporary file names, upstream introduced the max_file_uploads
  option in 5.3.1 and 5.2.12.

  Due to the nature of this new option a default limit has been set
  to 50, hoping it is sensible enough to not to cause disruptions on
  existing services.
  The value of this new limit can be changed in the php.ini file.

  If you installed the php5-suhosin extension there was a limiting
  mechanism in place already. In this case you may want to make sure
  the new limit imposed by PHP itself is not smaller than suhosin's.

 -- Raphael Geissert <geissert@debian.org>  Sat, 21 Nov 2009 13:37:51 -0600

php5 (5.2.6-1) unstable; urgency=medium

  * Now uses system timezone database.

  Debian PHP now makes use of the system wide timezone database from the
  tzdata package, making sure any updates there are automatically used
  by PHP aswell. Note that this requires that the PHP process has access
  to /etc/localtime and /usr/share/zoneinfo (this is usually the case).

  * New php5-dbg package.

  We are now shipping a php5-dgb package which will greatly aid in finding
  the cause of many crashes that you may experience.  So if you are going to
  report a bug for a reproducible crash, please install this package before
  sending a backtrace.

  * New libapache2-mod-php5filter package.

  We are now also shipping a new libapache2-mod-php5filter package which
  uses the "Apache 2.0 filter-module support via DSO through APXS".

 -- Thijs Kinkhorst <thijs@debian.org>  Wed, 23 Jul 2008 17:42:06 +0200

php5 (5.2.3-2) unstable; urgency=low

  The Suhosin patch is now enabled by default!

  For more information, see
  <http://www.hardened-php.net/suhosin/index.html>.

  Special thanks to Blars Blarson for providing a sparc machine for testing
  that the patch seems to work okay on that architecture.  If you experience
  otherwise let us know!

  Suggestions are welcome for default configuration options, examples,
  documentation, etc.

  In any event please report successes and/or failures to us at
  pkg-php-maint@lists.alioth.debian.org. 

 -- sean finney <seanius@debian.org>  Thu, 12 Jul 2007 23:38:43 +0200


ext/hash/hash_tiger.c
/*
  +----------------------------------------------------------------------+
  | PHP Version 5                                                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 1997-2012 The PHP Group                                |
  +----------------------------------------------------------------------+
  | This source file is subject to version 3.01 of the PHP license,      |
  | that is bundled with this package in the file LICENSE, and is        |
  | available through the world-wide-web at the following url:           |
  | http://www.php.net/license/3_01.txt                                  |
  | If you did not receive a copy of the PHP license and are unable to   |
  | obtain it through the world-wide-web, please send a note to          |
  | license@php.net so we can mail you a copy immediately.               |
  +----------------------------------------------------------------------+
  | Authors: Michael Wallner <mike@php.net>                              |
  |          Sara Golemon <pollita@php.net>                              |
  +----------------------------------------------------------------------+
*/

/* $Id: hash_tiger.c 321634 2012-01-01 13:15:04Z felipe $ */

#include "php_hash.h"
#include "php_hash_tiger.h"
#include "php_hash_tiger_tables.h"

#if (defined(__APPLE__) || defined(__APPLE_CC__)) && (defined(__BIG_ENDIAN__) || defined(__LITTLE_ENDIAN__))
# if defined(__LITTLE_ENDIAN__)
#  undef WORDS_BIGENDIAN
# else 
#  if defined(__BIG_ENDIAN__)
#   define WORDS_BIGENDIAN
#  endif
# endif
#endif

/* {{{ */
#define save_abc \
	aa = a; \
	bb = b; \
	cc = c;

#define round(a,b,c,x,mul) \
	c ^= x; \
	a -= t1[(unsigned char)(c)] ^ \
		t2[(unsigned char)(((php_hash_uint32)(c))>>(2*8))] ^ \
		t3[(unsigned char)((c)>>(4*8))] ^ \
		t4[(unsigned char)(((php_hash_uint32)((c)>>(4*8)))>>(2*8))] ; \
	b += t4[(unsigned char)(((php_hash_uint32)(c))>>(1*8))] ^ \
		t3[(unsigned char)(((php_hash_uint32)(c))>>(3*8))] ^ \
		t2[(unsigned char)(((php_hash_uint32)((c)>>(4*8)))>>(1*8))] ^ \
		t1[(unsigned char)(((php_hash_uint32)((c)>>(4*8)))>>(3*8))]; \
	b *= mul;

#define pass(a,b,c,mul) \
	round(a,b,c,x0,mul) \
	round(b,c,a,x1,mul) \
	round(c,a,b,x2,mul) \
	round(a,b,c,x3,mul) \
	round(b,c,a,x4,mul) \
	round(c,a,b,x5,mul) \
	round(a,b,c,x6,mul) \
	round(b,c,a,x7,mul)

#define key_schedule \
	x0 -= x7 ^ L64(0xA5A5A5A5A5A5A5A5); \
	x1 ^= x0; \
	x2 += x1; \
	x3 -= x2 ^ ((~x1)<<19); \
	x4 ^= x3; \
	x5 += x4; \
	x6 -= x5 ^ ((~x4)>>23); \
	x7 ^= x6; \
	x0 += x7; \
	x1 -= x0 ^ ((~x7)<<19); \
	x2 ^= x1; \
	x3 += x2; \
	x4 -= x3 ^ ((~x2)>>23); \
	x5 ^= x4; \
	x6 += x5; \
	x7 -= x6 ^ L64(0x0123456789ABCDEF);

#define feedforward \
	a ^= aa; \
	b -= bb; \
	c += cc;

#define compress(passes) \
	save_abc \
	pass(a,b,c,5) \
	key_schedule \
	pass(c,a,b,7) \
	key_schedule \
	pass(b,c,a,9) \
	for(pass_no=0; pass_no<passes; pass_no++) { \
		key_schedule \
		pass(a,b,c,9) \
		tmpa=a; a=c; c=b; b=tmpa; \
	} \
	feedforward

#define split_ex(str) \
	x0=str[0]; x1=str[1]; x2=str[2]; x3=str[3]; \
	x4=str[4]; x5=str[5]; x6=str[6]; x7=str[7];
#ifdef WORDS_BIGENDIAN
#	define split(str) \
	{ \
		int i; \
		php_hash_uint64 tmp[8]; \
		 \
		for (i = 0; i < 64; ++i) { \
			((unsigned char *) tmp)[i^7] = ((unsigned char *) str)[i]; \
		} \
		split_ex(tmp); \
	}
#else
#	define split split_ex
#endif

#define tiger_compress(passes, str, state) \
{ \
	register php_hash_uint64 a, b, c, tmpa, x0, x1, x2, x3, x4, x5, x6, x7; \
	php_hash_uint64 aa, bb, cc; \
	unsigned int pass_no; \
	\
	a = state[0]; \
	b = state[1]; \
	c = state[2]; \
	\
	split(str); \
	\
	compress(passes); \
	\
	state[0] = a; \
	state[1] = b; \
	state[2] = c; \
}
/* }}} */

static inline void TigerFinalize(PHP_TIGER_CTX *context)
{
	context->passed += (php_hash_uint64) context->length << 3;
	
	context->buffer[context->length++] = 0x1;
	if (context->length % 8) {
		memset(&context->buffer[context->length], 0, 8-context->length%8);
		context->length += 8-context->length%8;
	}
	
	if (context->length > 56) {
		memset(&context->buffer[context->length], 0, 64 - context->length);
		tiger_compress(context->passes, ((php_hash_uint64 *) context->buffer), context->state);
		memset(context->buffer, 0, 56);
	} else {
		memset(&context->buffer[context->length], 0, 56 - context->length);
	}

#ifndef WORDS_BIGENDIAN	
	memcpy(&context->buffer[56], &context->passed, sizeof(php_hash_uint64));
#else
	context->buffer[56] = (unsigned char) (context->passed & 0xff);
	context->buffer[57] = (unsigned char) ((context->passed >> 8) & 0xff);
	context->buffer[58] = (unsigned char) ((context->passed >> 16) & 0xff);
	context->buffer[59] = (unsigned char) ((context->passed >> 24) & 0xff);
	context->buffer[60] = (unsigned char) ((context->passed >> 32) & 0xff);
	context->buffer[61] = (unsigned char) ((context->passed >> 40) & 0xff);
	context->buffer[62] = (unsigned char) ((context->passed >> 48) & 0xff);
	context->buffer[63] = (unsigned char) ((context->passed >> 56) & 0xff);
#endif
	tiger_compress(context->passes, ((php_hash_uint64 *) context->buffer), context->state);
}

PHP_HASH_API void PHP_3TIGERInit(PHP_TIGER_CTX *context)
{
	memset(context, 0, sizeof(*context));
	context->state[0] = L64(0x0123456789ABCDEF);
	context->state[1] = L64(0xFEDCBA9876543210);
	context->state[2] = L64(0xF096A5B4C3B2E187);
}

PHP_HASH_API void PHP_4TIGERInit(PHP_TIGER_CTX *context)
{
	memset(context, 0, sizeof(*context));
	context->passes = 1;
	context->state[0] = L64(0x0123456789ABCDEF);
	context->state[1] = L64(0xFEDCBA9876543210);
	context->state[2] = L64(0xF096A5B4C3B2E187);
}

PHP_HASH_API void PHP_TIGERUpdate(PHP_TIGER_CTX *context, const unsigned char *input, size_t len)
{
	if (context->length + len < 64) {
		memcpy(&context->buffer[context->length], input, len);
		context->length += len;
	} else {
		size_t i = 0, r = (context->length + len) % 64;
		
		if (context->length) {
			i = 64 - context->length;
			memcpy(&context->buffer[context->length], input, i);
			tiger_compress(context->passes, ((const php_hash_uint64 *) context->buffer), context->state);
			memset(context->buffer, 0, 64);
			context->passed += 512;
		}
		
		for (; i + 64 <= len; i += 64) {
			memcpy(context->buffer, &input[i], 64);
			tiger_compress(context->passes, ((const php_hash_uint64 *) context->buffer), context->state);
			context->passed += 512;
		}
		memset(&context->buffer[r], 0, 64-r);
		memcpy(context->buffer, &input[i], r);
		context->length = r;
	}
}

PHP_HASH_API void PHP_TIGER128Final(unsigned char digest[16], PHP_TIGER_CTX *context)
{
	TigerFinalize(context);
	
	digest[0] = (unsigned char) ((context->state[0] >> 56) & 0xff);
	digest[1] = (unsigned char) ((context->state[0] >> 48) & 0xff);
	digest[2] = (unsigned char) ((context->state[0] >> 40) & 0xff);
	digest[3] = (unsigned char) ((context->state[0] >> 32) & 0xff);
	digest[4] = (unsigned char) ((context->state[0] >> 24) & 0xff);
	digest[5] = (unsigned char) ((context->state[0] >> 16) & 0xff);
	digest[6] = (unsigned char) ((context->state[0] >> 8) & 0xff);
	digest[7] = (unsigned char) (context->state[0] & 0xff);
	digest[8] = (unsigned char) ((context->state[1] >> 56) & 0xff);
	digest[9] = (unsigned char) ((context->state[1] >> 48) & 0xff);
	digest[10] = (unsigned char) ((context->state[1] >> 40) & 0xff);
	digest[11] = (unsigned char) ((context->state[1] >> 32) & 0xff);
	digest[12] = (unsigned char) ((context->state[1] >> 24) & 0xff);
	digest[13] = (unsigned char) ((context->state[1] >> 16) & 0xff);
	digest[14] = (unsigned char) ((context->state[1] >> 8) & 0xff);
	digest[15] = (unsigned char) (context->state[1] & 0xff);
	
	memset(context, 0, sizeof(*context));
}

PHP_HASH_API void PHP_TIGER160Final(unsigned char digest[20], PHP_TIGER_CTX *context)
{
	TigerFinalize(context);
	
	digest[0] = (unsigned char) ((context->state[0] >> 56) & 0xff);
	digest[1] = (unsigned char) ((context->state[0] >> 48) & 0xff);
	digest[2] = (unsigned char) ((context->state[0] >> 40) & 0xff);
	digest[3] = (unsigned char) ((context->state[0] >> 32) & 0xff);
	digest[4] = (unsigned char) ((context->state[0] >> 24) & 0xff);
	digest[5] = (unsigned char) ((context->state[0] >> 16) & 0xff);
	digest[6] = (unsigned char) ((context->state[0] >> 8) & 0xff);
	digest[7] = (unsigned char) (context->state[0] & 0xff);
	digest[8] = (unsigned char) ((context->state[1] >> 56) & 0xff);
	digest[9] = (unsigned char) ((context->state[1] >> 48) & 0xff);
	digest[10] = (unsigned char) ((context->state[1] >> 40) & 0xff);
	digest[11] = (unsigned char) ((context->state[1] >> 32) & 0xff);
	digest[12] = (unsigned char) ((context->state[1] >> 24) & 0xff);
	digest[13] = (unsigned char) ((context->state[1] >> 16) & 0xff);
	digest[14] = (unsigned char) ((context->state[1] >> 8) & 0xff);
	digest[15] = (unsigned char) (context->state[1] & 0xff);
	digest[16] = (unsigned char) ((context->state[2] >> 56) & 0xff);
	digest[17] = (unsigned char) ((context->state[2] >> 48) & 0xff);
	digest[18] = (unsigned char) ((context->state[2] >> 40) & 0xff);
	digest[19] = (unsigned char) ((context->state[2] >> 32) & 0xff);
	
	memset(context, 0, sizeof(*context));
}

PHP_HASH_API void PHP_TIGER192Final(unsigned char digest[24], PHP_TIGER_CTX *context)
{
	TigerFinalize(context);
	
	digest[0] = (unsigned char) ((context->state[0] >> 56) & 0xff);
	digest[1] = (unsigned char) ((context->state[0] >> 48) & 0xff);
	digest[2] = (unsigned char) ((context->state[0] >> 40) & 0xff);
	digest[3] = (unsigned char) ((context->state[0] >> 32) & 0xff);
	digest[4] = (unsigned char) ((context->state[0] >> 24) & 0xff);
	digest[5] = (unsigned char) ((context->state[0] >> 16) & 0xff);
	digest[6] = (unsigned char) ((context->state[0] >> 8) & 0xff);
	digest[7] = (unsigned char) (context->state[0] & 0xff);
	digest[8] = (unsigned char) ((context->state[1] >> 56) & 0xff);
	digest[9] = (unsigned char) ((context->state[1] >> 48) & 0xff);
	digest[10] = (unsigned char) ((context->state[1] >> 40) & 0xff);
	digest[11] = (unsigned char) ((context->state[1] >> 32) & 0xff);
	digest[12] = (unsigned char) ((context->state[1] >> 24) & 0xff);
	digest[13] = (unsigned char) ((context->state[1] >> 16) & 0xff);
	digest[14] = (unsigned char) ((context->state[1] >> 8) & 0xff);
	digest[15] = (unsigned char) (context->state[1] & 0xff);
	digest[16] = (unsigned char) ((context->state[2] >> 56) & 0xff);
	digest[17] = (unsigned char) ((context->state[2] >> 48) & 0xff);
	digest[18] = (unsigned char) ((context->state[2] >> 40) & 0xff);
	digest[19] = (unsigned char) ((context->state[2] >> 32) & 0xff);
	digest[20] = (unsigned char) ((context->state[2] >> 24) & 0xff);
	digest[21] = (unsigned char) ((context->state[2] >> 16) & 0xff);
	digest[22] = (unsigned char) ((context->state[2] >> 8) & 0xff);
	digest[23] = (unsigned char) (context->state[2] & 0xff);
	
	memset(context, 0, sizeof(*context));
}

#define PHP_HASH_TIGER_OPS(p, b) \
	const php_hash_ops php_hash_##p##tiger##b##_ops = { \
		(php_hash_init_func_t) PHP_##p##TIGERInit, \
		(php_hash_update_func_t) PHP_TIGERUpdate, \
		(php_hash_final_func_t) PHP_TIGER##b##Final, \
		(php_hash_copy_func_t) php_hash_copy, \
		b/8, \
		64, \
		sizeof(PHP_TIGER_CTX) \
	}

PHP_HASH_TIGER_OPS(3, 128);
PHP_HASH_TIGER_OPS(3, 160);
PHP_HASH_TIGER_OPS(3, 192);
PHP_HASH_TIGER_OPS(4, 128);
PHP_HASH_TIGER_OPS(4, 160);
PHP_HASH_TIGER_OPS(4, 192);

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: sw=4 ts=4 fdm=marker
 * vim<600: sw=4 ts=4
 */
